XML file based routing

SzymonPobiega opened the pull request:
This pull refactors the existing file-based routing to be more production ready. In order to do it properly I've redesigned the `Endpoint` class, replacing the `transport discriminator` with a generic property bag. The usage is following

```
var i = new EndpointInstance("Sales").SetProperty("schema", "nsb");
```

Transports might provide specific extension methods:

```
//MSMQ queue@machine
var i = new EndpointInstance("Sales").AtMachine("MyMachine");
```

The bag in future might support other properties (e.g. SQL Server's `schema` or gateway's `site`).

Other changes:
 * Removed exception when trying to scale out without providing user or transport discriminator. This is because the bag that replaced transport discriminator not necessarily guarantees name uniqueness.
 * Renamed `UniqueQueuePerEndpointInstance` to `UniqueInstanceDiscriminator`

### XML file based routing

The sample XML file is following:

```
<endpoints>
    <endpoint name="A">
        <instance discriminator="D1" prop1="V1" prop2="V2"/>
        <instance prop3="V3" prop4="V4"/>
    </endpoint>
    <endpoint name="B">
        <instance discriminator="D2" prop5="V5" prop6="V6"/>
    </endpoint>
</endpoints>
```

The only mandatory attribute is endpoint `name`. Parsing is lax, using DOM. The file based routing table checks the file every couple of seconds (removed file system watcher). If the file does not exist when starting up, the endpoint fails to start. If the file is deleted during run-time, errors are logged but endpoint is not killed.

SzymonPobiega commented:
@janovesk @tmasternak  please review.

tmasternak commented on src/NServiceBus.Core/EndpointInstance.cs:
```diff
@@ -1,21 +1,33 @@
+        /// <summary>
+        /// A specific discriminator for scale-out purposes.
+        /// </summary>
+        public string ScaleOutDiscriminator { get; }
```
I we are representing scale out concepts in EndpointInstance maybe we should pull here qualifer from LocalAddress and call it SatelliteDiscriminator (which might be more expressive)?

SzymonPobiega replied:
Not sure about committing to make satellites fully supported concepts. I am on the fence here. On the other hand, no matter how the current implementation of the satellites is broken, the concept itself makes sense and I see there will be need for privately-scoped "sub addresses"...

tmasternak commented on src/NServiceBus.Core/Transports/TransportDefinition.cs:
```diff
@@ -43,7 +43,7 @@ public abstract partial class TransportDefinition
         /// Returns the discriminator for this endpoint instance.
         /// </summary>
-        public abstract string GetDiscriminatorForThisEndpointInstance(ReadOnlySettings settings);
+        public abstract EndpointInstance BindToLocalEndpoint(EndpointInstance instance, ReadOnlySettings settings);
```
What about calling this method AdjustLocalEndpointInstance?

SzymonPobiega replied:
I don't think it conveys the same intent which is "return an endpoint instance which is specific for this local endpoint", hence "bind".

tmasternak replied:
:+1:

tmasternak commented:
Why not using a composition over BCL Dictionary in PropertyDictionary instead of custom list implementation?

tmasternak commented on src/NServiceBus.Core/Routing/FileBasedDynamicRouting/FileRoutingTable.cs:
```diff
@@ -1,189 +1,109 @@
         }
 
-        void StartMonitoring(string basePath, string queueName, string fileName)
+        async Task ReloadData()
```
Currently we have two level retries. Immediate retries done inside ReloadData and loop inside AsyncTimer do we want both? Should we enable endpoint instance to start even if we did not load a routing file a single time?

SzymonPobiega replied:
I think two level makes sense. For production use we want to have the first level refresh be done relatively rarely (one a minute?) but if we fail, we want to retry relatively frequently (every second) since the errors are probably caused by somebody updating the file.

Not sure about that. If user enables the file-based routing, he expresses the intent to use that file. And if it is not present that means that probably most (if not all) of routing is not configured so the endpoint is not functional.

tmasternak commented on src/NServiceBus.Core/Routing/FileBasedDynamicRouting/FileRoutingTableParser.cs:
```diff
@@ -0,0 +1,38 @@
+                var endpointName = new Endpoint(nameAttribute.Value);
+                foreach (var i in e.Descendants("instance"))
+                {
+                    var discriminatorAttribute = i.Attribute("discriminator");
```
Should we changed the attribute name to scaleOutDiscriminator to express it's current usage?

SzymonPobiega replied:
:+1:

SzymonPobiega commented:
Rebased.

andreasohlund commented on src/NServiceBus.AcceptanceTests/Msmq/When_not_providing_user_discriminator.cs:
```diff
@@ -8,10 +8,10 @@
     {
         [Test]
-        public async Task Should_be_a_no_op_discriminator()
+        public async Task Should_use_only_the_machine_name()
```
So we now default to machine name for all transports/hosting models?

andreasohlund replied:
Strike that, saw that this still was a msmq specific test

SzymonPobiega commented:
Hold your horses, does not compile;)

andreasohlund commented on src/NServiceBus.AcceptanceTests/Routing/When_broadcasting_a_command.cs:
```diff
@@ -34,17 +34,18 @@ public class Sender : EndpointConfigurationBuilder
+                var filePath = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "routes.xml");
+                File.WriteAllText(filePath, @"<endpoints>
+    <endpoint name=""DistributingACommand.Receiver"">
+        <instance scaleOutDiscriminator=""1""/>
```
does `scaleOut` add any value? (we still need to explain to them what a discriminator is?) rename to `discriminator` ?

SzymonPobiega replied:
@tmasternak ? You preferred `scaleOut` prefix

tmasternak replied:
I like ~~scaleOut. This's used only for that and it's more expressive than just ~~discriminator

andreasohlund replied:
>I like ~~scaleOut. This's used only for that and it's more expressive than just ~~discriminator

Not sure I agree, instances and how to discriminate them is by definition all about scaling out so I don't see what `scaleOut` adds

SzymonPobiega replied:
OK guys, I am on the fence here:
 * @andreasohlund wants `discriminator`
 * @tmasternak wants `scaleOutDiscriminator`
 * @timbussmann wants `instanceId`
I'll change it back to discriminator not because I prefer what @andreasohlund says but because this is what it was called before. If you guys feel like it needs to be changed, please propose another PR. I don't want this rename to block the whole PR.

andreasohlund replied:
:+1:

andreasohlund commented on src/NServiceBus.AcceptanceTests/Routing/When_broadcasting_a_command.cs:
```diff
@@ -77,7 +78,7 @@ public Receiver1()
                 {
                     c.EndpointName("DistributingACommand.Receiver");
-                    c.ScaleOut().UniqueQueuePerEndpointInstance("1");
+                    c.ScaleOut().UniqueInstanceDiscriminator("1");
```
Does `Unique` add any value? What about `InstanceDiscriminator` or `EndpointInstanceDiscriminator`  ?

SzymonPobiega replied:
:+1: for `InstanceDiscriminator`

andreasohlund commented on src/NServiceBus.Core.Tests/API/APIApprovals.ApproveNServiceBus.approved.txt:
```diff
@@ -306,15 +306,20 @@ namespace NServiceBus
-        public string TransportDiscriminator { get; }
-        public string UserDiscriminator { get; }
+        public System.Collections.Generic.IReadOnlyDictionary<string, string> Properties { get; }
+        public string ScaleOutDiscriminator { get; }
```
Rename to just `Discriminator`?

andreasohlund commented on src/NServiceBus.Core.Tests/API/APIApprovals.ApproveNServiceBus.approved.txt:
```diff
@@ -306,15 +306,20 @@ namespace NServiceBus
         public NServiceBus.Endpoint Endpoint { get; }
-        public string TransportDiscriminator { get; }
-        public string UserDiscriminator { get; }
+        public System.Collections.Generic.IReadOnlyDictionary<string, string> Properties { get; }
```
Is the user supposed to set properties or can we make this internal?

SzymonPobiega replied:
They should be able to set but probably not get...

andreasohlund commented on src/NServiceBus.Core/Transports/Msmq/EndpointInstanceExtensions.cs:
```diff
@@ -0,0 +1,19 @@
+        /// </summary>
+        /// <param name="instance">A plain instance.</param>
+        /// <param name="machineName">Machine name.</param>
+        public static EndpointInstance AtMachine(this EndpointInstance instance, string machineName)
```
Couldn't find any tests for this option?

andreasohlund commented on src/NServiceBus.Core.Tests/Routing/EndpointInstancesTests.cs:
```diff
@@ -12,7 +12,7 @@ public class EndpointInstancesTests
         public void Should_throw_when_trying_to_configure_instances_that_don_not_match_endpoint_name()
```
Nitpick: Typo don_not

SzymonPobiega replied:
:+1:

andreasohlund commented on src/NServiceBus.Core/EndpointInstance.cs:
```diff
@@ -1,21 +1,33 @@
-        /// <param name="transportDiscriminator">The discriminator provided by the transport, if any.</param>
-        public EndpointInstance(string endpoint, string userDiscriminator, string transportDiscriminator)
-            : this(new Endpoint(endpoint), userDiscriminator, transportDiscriminator)
+        /// <param name="scaleOutDiscriminator">A specific discriminator for scale-out purposeses.</param>
```
typo

SzymonPobiega replied:
:+1:

andreasohlund commented on src/NServiceBus.Core/EndpointInstance.cs:
```diff
@@ -23,38 +35,39 @@ public EndpointInstance(string endpoint, string userDiscriminator, string transp
-        /// <param name="userDiscriminator">The discriminator provided by the user, if any.</param>
-        /// <param name="transportDiscriminator">The discriminator provided by the transport, if any.</param>
-        public EndpointInstance(Endpoint endpoint, string userDiscriminator, string transportDiscriminator)
+        /// <param name="scaleOutDiscriminator">A specific discriminator for scale-out purposeses.</param>
```
typo

SzymonPobiega replied:
:+1:

andreasohlund commented on src/NServiceBus.Core/Routing/FileBasedDynamicRouting/AsyncTimer.cs:
```diff
@@ -0,0 +1,44 @@
+                {
+                    try
+                    {
+                        await Task.Delay(interval, token);
```
This will ++ the @danielmarbach :beer: count

SzymonPobiega replied:
:+1:

andreasohlund commented on src/NServiceBus.Core/Routing/FileBasedDynamicRouting/FileRoutingTable.cs:
```diff
@@ -1,189 +1,109 @@
-        }
+            var doc = await ReadFileWithRetries().ConfigureAwait(false);
+            var instances = parser.Parse(doc);
+            var newInsatnceMap = new Dictionary<Endpoint, HashSet<EndpointInstance>>();
```
typo

SzymonPobiega replied:
:+1:

andreasohlund commented on src/NServiceBus.Core/Routing/FileBasedDynamicRouting/FileRoutingTable.cs:
```diff
@@ -1,189 +1,109 @@
+                        throw;
                     }
                 }
+                await Task.Delay(TimeSpan.FromSeconds(1));
```
more :beers: to @danielmarbach

SzymonPobiega replied:
:+1:

andreasohlund commented on src/NServiceBus.Core/obsoletes.cs:
```diff
@@ -1012,11 +1012,29 @@ public void UseSingleBrokerQueue()
+        [ObsoleteEx(
+            TreatAsErrorFromVersion = "6",
+            RemoveInVersion = "7",
+            Message = "Not required any more as for MSMQ that behavior is default and for other transports the unique instance ID has be provided.")]
```
Typo: has to BE provided

SzymonPobiega replied:
:+1:

andreasohlund commented on src/NServiceBus.Core/obsoletes.cs:
```diff
@@ -1012,11 +1012,29 @@ public void UseSingleBrokerQueue()
             TreatAsErrorFromVersion = "6",
             RemoveInVersion = "7",
-            ReplacementTypeOrMember = "UniqueQueuePerEndpointInstance")]
+            Message = "Not required any more as for MSMQ that behavior is default and for other transports the unique instance ID has be provided.")]
```
Typo: has to BE provided

SzymonPobiega commented:
Thanks a lot @andreasohlund ! I fixed everything except `scaleOutDiscriminator` vs `discriminator`

danielmarbach commented on src/NServiceBus.Core/Routing/FileBasedDynamicRouting/RoutingFileAccess.cs:
```diff
@@ -0,0 +1,23 @@
+        public Task<XDocument> Load(string path)
+        {
+            XDocument doc;
+            using (var file = new FileStream(path, FileMode.Open, FileAccess.Read, FileShare.ReadWrite))
```
For my on education I googled XDocument and async. The only thing I found was this

```
using (XmlReader r = XmlReader.Create(new StringReader(input), new XmlReaderSettings() { Async = true }))
            {
                while (await r.ReadAsync())
                {
                    switch (r.NodeType)
                    {
                        case XmlNodeType.Element:
                            Console.WriteLine(r.LocalName);
                            break;

                        case XmlNodeType.Text:
                            Console.WriteLine(await r.GetValueAsync());
                            break;
                    }
                }
            }
```

SzymonPobiega replied:
That does not look like `XDocument` ;-) I could not find any `async` overload of the `XDocument.Load` and I am not sure if asynchronously loading a file into a string and then calling `XDocument.Parse` will be better.

timbussmann commented on src/NServiceBus.Core/PropertyDictionary.cs:
```diff
@@ -0,0 +1,173 @@
+    using System.Collections;
+    using System.Collections.Generic;
+
+    sealed class PropertyDictionary : IReadOnlyDictionary<string, string>
```
why do we need a custom "linked dictionary" implementation?

SzymonPobiega replied:
I wanted to make `EndpointInstance` a read-only and copying stuff from one dictionary to another seemed like a significant waste.

timbussmann replied:
while the implementation looks valid to me, I'd strongly prefer to find a simpler solution and not relying on a custom collection.

timbussmann commented on src/NServiceBus.Core/Routing/FileBasedDynamicRouting/FileRoutingTable.cs:
```diff
@@ -1,189 +1,110 @@
-                }
-            }
-        }
+            await ReloadData();
```
:beer: ++

SzymonPobiega replied:
:+1:

timbussmann commented on src/NServiceBus.Core/Routing/FileBasedDynamicRouting/FileRoutingTableFeature.cs:
```diff
@@ -1,46 +1,32 @@
 
     class FileRoutingTableFeature : Feature
     {
+        public static readonly string CheckIntervalSettingsKey = "FileBasedRouting.CheckInterval";
```
remove readonly. Could you use `const` in this case?

SzymonPobiega replied:
:+1:

timbussmann commented on src/NServiceBus.Core/Routing/FileBasedDynamicRouting/RoutingFileAccess.cs:
```diff
@@ -0,0 +1,23 @@
+                    doc = XDocument.Load(reader);
+                }
+            }
+            return Task.FromResult(doc);
```
do we need an async (internal) interface which calls no async operations?

SzymonPobiega replied:
:+1:

timbussmann commented on src/NServiceBus.Core/obsoletes.cs:
```diff
@@ -1012,11 +1012,29 @@ public void UseSingleBrokerQueue()
             TreatAsErrorFromVersion = "6",
             RemoveInVersion = "7",
-            ReplacementTypeOrMember = "UniqueQueuePerEndpointInstance")]
+            Message = "Not required any more as for MSMQ that behavior is default and for other transports the unique instance ID has to be provided.")]
```
maybe "[...] as for MSMQ this is the default behavior [...]"

SzymonPobiega replied:
:+1:

timbussmann commented:
I really dislike all these "discriminators" in the namings/configurations/parameters. Couldn't we use something like "instance identifier"?

timbussmann commented:
Also, should we move `EndpointInstance` and related classes into a `Routing` namespace since this is easily messed up with classes like `RunningEndpointInstance` which is something entirely different but in the same namespace.

timbussmann commented on src/NServiceBus.Core/EndpointInstance.cs:
```diff
@@ -23,38 +35,39 @@ public EndpointInstance(string endpoint, string userDiscriminator, string transp
-                throw new ArgumentException("Transport-provided discriminator cannot be an empty string.");
+                foreach (var kvp in properties)
+                {
+                    this.properties = this.properties.Set(kvp.Key, kvp.Value);
```
can't we just store the ReadOnlyDictionary?

SzymonPobiega commented:
@timbussmann Implemented everything but the dictionary change. I'll try to spike it now to see if I like it. Thank!

SzymonPobiega commented:
I addressed the rest of the comments:
 * Back to `Discriminator` name
 * Removed custom endpoint instance property dictionary
 * Moved `Endpoint` and `EndpointInstance` to routing namespace. 

@tmasternak @andreasohlund @timbussmann please merge when green.

timbussmann commented:
Good with merging that :+1: @SzymonPobiega do you want to squash first?

SzymonPobiega commented:
Squashed. I'll merge when green.

timbussmann commented:
are you going to update/add doco for that?

SzymonPobiega commented:
Doing it now but I detected that moving the `Endpoint` to routing broke the factory methods so now trying to fix it by splitting core endpoint from routing endpoint into two separate classes.
